File: java/MergeSort.java
import java.io.*;
import java.util.*;

class MergeSort {
    static Scanner sc = new Scanner(System.in);
    public static ArrayList<Integer> mergeSort(ArrayList<Integer> list) {
        if (list.size() <= 1) return list;
        else return merge(
            mergeSort(new ArrayList<Integer>(list.subList(0, list.size() / 2))),
            mergeSort(new ArrayList<Integer>(list.subList(list.size() / 2, list.size())))
        );
    }
    public static ArrayList<Integer> merge(ArrayList<Integer> arr1, ArrayList<Integer> arr2) {
        ArrayList<Integer> merged = new ArrayList<>();
        while (arr1.size() > 0 && arr2.size() > 0) {
            if (arr1.get(0) < arr2.get(0)) merged.add(arr1.remove(0));
            else if (arr1.get(0) > arr2.get(0)) merged.add(arr2.remove(0));
            else {
                merged.add(arr1.remove(0));
                merged.add(arr2.remove(0));
            }
        }
        merged.addAll(arr1);
        merged.addAll(arr2);
        return merged;
    }
    public static ArrayList<Integer> readFromFile() {
        try {
            ArrayList<Integer> arr = new ArrayList<>();
            System.out.println("Enter the file name (Make sure each elements are in new line): ");
            String fileName = sc.next();
            Scanner sfile = new Scanner(new File(fileName));
            while (sfile.hasNextInt()) arr.add(sfile.nextInt());
            return arr;
        } catch (Exception e) {
            System.out.println(e);
        }
        return null;
    }
    public static ArrayList<Integer> generateRandomArray() {
        System.out.println("Enter the size of array : ");
        Random r = new Random();
        int size = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        for (int i = 0; i < size; i++) arr.add(r.nextInt(1000000));
        return arr;
    }
    public static void printArray(ArrayList<Integer> arr) {
        for (int i : arr) System.out.print(i + " ");
        System.out.println();
    }
    public static void main(String[] args) {
        System.out.println("Enter your choice :\n1.Read from file\n2.Randomly generated array ");
        int choice = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        switch (choice) {
            case 1:
                arr = readFromFile();
                break;
            case 2:
                arr = generateRandomArray();
                break;
            default:
                System.out.println("Invalid choice");
                break;
        }
        System.out.println("The array is  : ");
        printArray(arr);
        System.out.println("Sorted array is : ");
        long start = System.nanoTime();
        arr = mergeSort(arr);
        long end = System.nanoTime();
        printArray(arr);
        System.out.println("Time taken : " + (end - start) / 1000 + " microseconds");
    }
}

File: Python/mergesort.py
import time
import random

def mergesort(uarr: list) -> list :
    if len(uarr) <= 1:
        return uarr;
    mid: int = len(uarr)//2
    return merge(mergesort(uarr[:mid]), mergesort(uarr[mid:]))

def merge(arr1: list, arr2: list) -> list :
    i: int = 0
    j: int = 0
    merged: list = []

    while (j < len(arr2)) and (i < len(arr1)) :
        if arr1[i] == arr2[j] :
            merged = merged + [arr1[i],arr2[j]]
            j+=1
            i+=1
        elif arr1[i] < arr2[j] :
            merged = merged + [arr1[i]]
            i+=1
        else:
            merged = merged + [arr2[j]]
            j+=1
    if i < len(arr1) :
        merged = merged + arr1[i:]
    if j < len(arr2) :
        merged = merged + arr2[j:]
    
    return merged


def print_arr(arr):
    for i in arr:
        print(i, end=" ")
    print()


ch = int(input(
    "Enter your choice:\n1.Read from file\n2.Generate random array\n=>"
))
if ch == 1:
    filename = input("Enter file name: ")
    with open(filename, 'r') as f:
        arr = [int(x) for x in f.read().split()]
elif ch == 2:
    n = int(input("Enter number of elements: "))
    arr = [random.randint(0, 100000000) for _ in range(n)]
else:
    print("Invalid choice")

print("The array is : ")
print_arr(arr)
start = time.time()
sorted_arr = mergesort(arr)
end = time.time()
print("The sorted array is : ")
print_arr(sorted_arr)

print("The timetaken to sort array of {} elements is {} ms"
      .format(n, (end - start)*1000))
